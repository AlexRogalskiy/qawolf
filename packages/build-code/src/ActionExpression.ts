export class ActionExpression {
  code() {
    return "";
  }

  updatableCode() {
    return this.code();
  }
}
