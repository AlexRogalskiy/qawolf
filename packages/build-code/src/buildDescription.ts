import { Step } from "@qawolf/types";
import { describeDoc } from "@qawolf/web";
import { isNil } from "lodash";

export const buildDescription = (step: Step): string => {
  if (step.action === "scroll") {
    return `scroll`;
  }

  const description = describeDoc(step.html);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `select${description}`;
  }

  if (step.action === "type") {
    if (step.replace && (step.value === "" || isNil(step.value))) {
      return `clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value && value.indexOf("↓Enter") === 0) {
      return `Enter`;
    }

    if (value && value.indexOf("↓Tab") === 0) {
      return `Tab`;
    }

    return `type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
