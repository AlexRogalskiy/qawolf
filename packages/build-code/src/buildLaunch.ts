export const buildLaunch = (url: string, device?: string): string => {
  if (device) {
    return `await launch({ device: "${device}", url: "${url}" });`;
  }

  return `await launch({ url: "${url}" });`;
};
