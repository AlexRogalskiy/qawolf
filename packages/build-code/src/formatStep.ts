import { Step } from "@qawolf/types";
import { formatIt } from "./formatIt";
import { formatMethod } from "./formatMethod";

export const formatStep = (step: Step) => {
  return {
    it: formatIt(step),
    method: formatMethod(step.action, step.index, step.value)
  };
};
