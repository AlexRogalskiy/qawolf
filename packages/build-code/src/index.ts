export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
