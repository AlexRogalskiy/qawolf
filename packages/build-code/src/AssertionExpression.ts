import { ActionExpression } from "./ActionExpression";

type ConstructorOptions = {
  action: ActionExpression;
  description: string;
};

export class AssertionExpression {
  private _action: ActionExpression;
  private _description: string;

  constructor({ action, description }: ConstructorOptions) {
    this._action = action;
    this._description = description;
  }

  private _formatIt() {
    const description = this._description;
    if (!description.includes(`"`)) return `"can ${description}"`;
    if (!description.includes("'")) return `'can ${description}'`;
    return JSON.stringify("can " + description);
  }

  code() {
    const code = `\nit(${this._formatIt()}, async () => {\n  ${this._action.code()}\n  });\n`;
    return code;
  }

  updatableCode() {
    return this._action.updatableCode();
  }
}
