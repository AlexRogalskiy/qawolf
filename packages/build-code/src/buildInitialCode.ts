import { camelCase } from "lodash";
import { buildLaunch } from "./buildLaunch";

export type InitialCodeOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  patchHandle: string;
  url: string;
};

type TemplateOptions = {
  launch: string;
  name: string;
  patchHandle: string;
};

const buildInitialScript = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

const ${name} = async browser => {
  ${patchHandle}
};

exports.${name} = ${name};

if (require.main === module) {
  (async () => {
    const browser = ${launch}
    await ${name}(browser);
    await browser.close();
  })();
}`;

  return code;
};

const buildInitialTest = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

describe('${name}', () => {
  let browser;

  beforeAll(async () => {
    browser = ${launch}
  });

  afterAll(() => browser.close());
  ${patchHandle}
});`;

  return code;
};

export const buildInitialCode = (options: InitialCodeOptions) => {
  const templateOptions = {
    launch: buildLaunch(options.url, options.device),
    name: camelCase(options.name),
    patchHandle: options.patchHandle
  };

  if (options.isTest) {
    return buildInitialTest(templateOptions);
  }

  return buildInitialScript(templateOptions);
};
