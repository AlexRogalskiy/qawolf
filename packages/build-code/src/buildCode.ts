import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { camelCase } from "lodash";
import { resolve } from "path";
import { formatLaunch } from "./formatLaunch";
import { formatStep } from "./formatStep";

type BuildCodeOptions = {
  test?: boolean;
  workflow: Workflow;
};

const scriptTemplate = compile(
  readFileSync(resolve(__dirname, "../static/script.hbs"), "utf8")
);

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildCode = (options: BuildCodeOptions) => {
  const { workflow } = options;

  const template = options.test ? testTemplate : scriptTemplate;

  const code = template({
    launch: formatLaunch(workflow.url, workflow.device),
    name: camelCase(workflow.name),
    steps: workflow.steps.map((step, i) => {
      const previousStep = i > 0 ? workflow.steps[i - 1] : null;

      return formatStep(step, previousStep);
    }),
    url: workflow.url
  });

  return code;
};
