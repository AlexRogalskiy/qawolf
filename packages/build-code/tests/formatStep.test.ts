import { Step } from "@qawolf/types";
import { htmlToDoc } from "@qawolf/web";
import { formatStep } from "../src/formatStep";

const doc = htmlToDoc;

describe("formatStep", () => {
  describe("click", () => {
    it("formats click on link step", () => {
      const step: Step = {
        action: "click",
        html: {
          ancestors: [],
          node: doc("<a innertext='contact'>contact</a>")
        },
        index: 0,
        isFinal: true,
        page: 0
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats click on submit input step", () => {
      const step: Step = {
        action: "click",
        html: {
          ancestors: [],
          node: doc(`<input innertext="someone's" />`)
        },
        index: 1,
        isFinal: true,
        page: 0
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("scroll", () => {
    it("formats scroll action", () => {
      const step: Step = {
        action: "scroll",
        html: {
          ancestors: [],
          node: doc("<html />")
        },
        index: 0,
        isFinal: true,
        page: 0,
        value: { x: 0, y: 10 }
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("select", () => {
    it("formats select action", () => {
      const step: Step = {
        action: "select",
        html: {
          ancestors: [],
          node: doc("<select name='select1' />")
        },
        index: 0,
        isFinal: true,
        page: 0,
        value: "spirit"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("type", () => {
    it("formats clear an input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: doc("<input />")
        },
        index: 0,
        isFinal: true,
        page: 0,
        value: null
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats type into text input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: doc(
            `<input id='input1' name='username' placeholder='Jane Doe' />`
          )
        },
        index: 0,
        isFinal: true,
        page: 0,
        value: "spirit"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats type into password input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: doc(`<input id='input2' placeholder='secret' />`)
        },
        index: 10,
        isFinal: true,
        page: 0,
        value: "supersecret"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });
});
