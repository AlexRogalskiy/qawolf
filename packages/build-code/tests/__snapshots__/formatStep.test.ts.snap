// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatStep click formats click on link step 1`] = `
Object {
  "it": "click \\"contact\\" link",
  "method": "await browser.click(selectors[0]);",
}
`;

exports[`formatStep click formats click on submit input step 1`] = `
Object {
  "it": "click \\"someones\\" input",
  "method": "await browser.click(selectors[1]);",
}
`;

exports[`formatStep scroll formats scroll action 1`] = `
Object {
  "it": "scroll",
  "method": "await browser.scroll(selectors[0], { x: 0, y: 10 });",
}
`;

exports[`formatStep select formats select action 1`] = `
Object {
  "it": "select \\"select1\\"",
  "method": "await browser.select(selectors[0], \\"spirit\\");",
}
`;

exports[`formatStep type formats clear an input 1`] = `
Object {
  "it": "clear input",
  "method": "await browser.type(selectors[0], null);",
}
`;

exports[`formatStep type formats type into password input 1`] = `
Object {
  "it": "type into \\"secret\\" input",
  "method": "await browser.type(selectors[10], \\"supersecret\\");",
}
`;

exports[`formatStep type formats type into text input 1`] = `
Object {
  "it": "type into \\"username\\" input",
  "method": "await browser.type(selectors[0], \\"spirit\\");",
}
`;
